JBang alias don't seem to resolve from custom/snapshot repositories

**Describe the bug**
I can not add an alias or run a alias that has configured custom repositories.

Trying to run this alias https://github.com/ZenWave360/zenwave-code-generator/blob/main/jbang-catalog.json 

```json
{
  "aliases": {
    "zw": {
      "script-ref": "io.github.zenwave360:zenwave-code-generator-cli:0.0.1-SNAPSHOT",
      "dependencies": [
        "io.github.zenwave360.zenwave-code-generator.plugins:asyncapi-spring-cloud-streams3:0.0.1-SNAPSHOT",
        "io.github.zenwave360.zenwave-code-generator.plugins:openapi-spring-webtestclient:0.0.1-SNAPSHOT",
        "io.github.zenwave360.zenwave-code-generator.plugins:jdl-entities:0.0.1-SNAPSHOT",
        "io.github.zenwave360.zenwave-code-generator.plugins:jdl-to-openapi:0.0.1-SNAPSHOT"
      ],
      "repositories": [
        "mavencentral",
        "snapshots=https://s01.oss.sonatype.org/content/repositories/snapshots"
      ],
      "main": "io.zenwave360.generator.Main"
    }
  }
}
```

**To Reproduce**
Steps to reproduce the behavior:

```shell
jbang zw@zenwave360/zenwave-code-generator

[jbang] [WARN] Detected missing or out-of-date dependencies in cache.
[jbang] Resolving dependencies...
[jbang] io.github.zenwave360:zenwave-code-generator-cli:jar:0.0.1-SNAPSHOT
[jbang] [ERROR] Could not resolve dependencies from mavencentral=https://repo1.maven.org/maven2/
Unable to collect/resolve dependency tree for a resolution due to: Could not find artifact io.github.zenwave360:zenwave-code-generator-cli:jar:0.0.1-SNAPSHOT in mavencentral (https://repo1.maven.org/maven2/), caused by: Could not find artifact io.github.zenwave360:zenwave-code-generator-cli:jar:0.0.1-SNAPSHOT in mavencentral (https://repo1.maven.org/maven2/)
  Could not find artifact io.github.zenwave360:zenwave-code-generator-cli:jar:0.0.1-SNAPSHOT in mavencentral (https://repo1.maven.org/maven2/)
```

**Expected behavior**
Being able to add an alias and/or run a remote catalog alias with custom repositories

**JBang version**
```shell
jbang version --verbose
[jbang] jbang version 0.94.0
Cache: C:\Users\ivan.garcia\.jbang\cache
Config: C:\Users\ivan.garcia\.jbang
Repository:C:\Users\ivan.garcia\.m2\repository
0.94.0
```

**Additional context**
Add any other context about the problem here.

--- a/src/main/java/dev/jbang/cli/Alias.java
+++ b/src/main/java/dev/jbang/cli/Alias.java
@@ -74,7 +74,7 @@ class AliasAdd extends BaseAliasCommand {
 
 	@CommandLine.Option(names = { "-m",
 			"--main" }, description = "Main class to use when running. Used primarily for running jar's.")
-	String mainClass;
+	String main;
 
 	@CommandLine.Option(names = { "-j",
 			"--java" }, description = "JDK version to use for running the alias.")
@@ -96,7 +96,7 @@ class AliasAdd extends BaseAliasCommand {
 					"Invalid alias name, it should start with a letter followed by 0 or more letters, digits, underscores or hyphens");
 		}
 
-		RunContext ctx = RunContext.empty();
+		RunContext ctx = getRunContext();
 		Code code = ctx.forResource(scriptOrFile);
 		if (name == null) {
 			name = CatalogUtil.nameFromRef(ctx.getOriginalRef());
@@ -108,14 +108,32 @@ class AliasAdd extends BaseAliasCommand {
 		if (catFile != null) {
 			CatalogUtil.addAlias(catFile, name, scriptOrFile, desc, userParams, javaRuntimeOptions, sources,
 					dependencyInfoMixin.getDependencies(), dependencyInfoMixin.getRepositories(),
-					dependencyInfoMixin.getClasspaths(), dependencyInfoMixin.getProperties(), javaVersion, mainClass);
+					dependencyInfoMixin.getClasspaths(), dependencyInfoMixin.getProperties(), javaVersion, main);
 		} else {
 			catFile = CatalogUtil.addNearestAlias(name, scriptOrFile, desc, userParams, javaRuntimeOptions, sources,
 					dependencyInfoMixin.getDependencies(), dependencyInfoMixin.getRepositories(),
-					dependencyInfoMixin.getClasspaths(), dependencyInfoMixin.getProperties(), javaVersion, mainClass);
+					dependencyInfoMixin.getClasspaths(), dependencyInfoMixin.getProperties(), javaVersion, main);
 		}
 		info(String.format("Alias '%s' added to '%s'", name, catFile));
 		return EXIT_OK;
+	}
+
+	RunContext getRunContext() {
+		RunContext ctx = new RunContext();
+		ctx.setProperties(dependencyInfoMixin.getProperties());
+		ctx.setAdditionalDependencies(dependencyInfoMixin.getDependencies());
+		ctx.setAdditionalRepositories(dependencyInfoMixin.getRepositories());
+		ctx.setAdditionalClasspaths(dependencyInfoMixin.getClasspaths());
+		ctx.setAdditionalSources(sources);
+		ctx.setJavaVersion(javaVersion);
+		ctx.setMainClass(main);
+		Path cat = getCatalog(false);
+		if (cat != null) {
+			ctx.setCatalog(cat.toFile());
+		}
+		ctx.setArguments(userParams);
+		ctx.setJavaOptions(javaRuntimeOptions);
+		return ctx;
 	}
 }
 

--- a/src/main/java/dev/jbang/source/ResourceResolver.java
+++ b/src/main/java/dev/jbang/source/ResourceResolver.java
@@ -1,9 +1,5 @@
 package dev.jbang.source;
 
-import java.util.function.Function;
-
-import dev.jbang.catalog.Catalog;
-import dev.jbang.dependencies.ModularClassPath;
 import dev.jbang.source.resolvers.*;
 
 /**
@@ -57,26 +53,6 @@ public interface ResourceResolver {
 	}
 
 	String description();
-
-	/**
-	 * Factory method that returns a resource resolver that knows how to deal with
-	 * script/source files. It should be passed a function for dealing with Maven
-	 * GAVs.
-	 *
-	 * @param depResolver A function which, given a GAV string, returns a
-	 *                    <code>ModularClassPath</code>
-	 * @return A <code>ResourceRef</code> or <code>null</code>
-	 */
-	static ResourceResolver forScripts(Function<String, ModularClassPath> depResolver, Catalog catalog) {
-		return new AliasResourceResolver(catalog,
-				new CombinedResourceResolver(
-						new RenamingScriptResourceResolver(),
-						new LiteralScriptResourceResolver(),
-						new RemoteResourceResolver(false),
-						new ClasspathResourceResolver(),
-						new GavResourceResolver(depResolver),
-						new FileResourceResolver()));
-	}
 
 	/**
 	 * Factory method that returns a resource resolver that knows how to deal with

--- a/src/main/java/dev/jbang/source/RunContext.java
+++ b/src/main/java/dev/jbang/source/RunContext.java
@@ -10,7 +10,7 @@ import dev.jbang.catalog.Catalog;
 import dev.jbang.cli.BaseCommand;
 import dev.jbang.cli.ExitException;
 import dev.jbang.dependencies.*;
-import dev.jbang.source.resolvers.AliasResourceResolver;
+import dev.jbang.source.resolvers.*;
 import dev.jbang.util.JavaUtil;
 import dev.jbang.util.PropertiesValueResolver;
 import dev.jbang.util.Util;
@@ -417,62 +417,17 @@ public class RunContext {
 	}
 
 	private List<Source> allToScriptSource(List<String> sources) {
-		Catalog catalog = getCatalog() != null ? Catalog.get(getCatalog().toPath()) : null;
-		ResourceResolver resolver = ResourceResolver.forScripts(this::resolveDependency, catalog);
 		Function<String, String> propsResolver = it -> PropertiesValueResolver.replaceProperties(it,
 				getContextProperties());
 		return sources	.stream()
 						.flatMap(line -> Util.explode(null, Util.getCwd(), line).stream())
-						.map(s -> resolveChecked(resolver, s))
+						.map(s -> resolveChecked(getResourceResolver(), s))
 						.map(ref -> Source.forResourceRef(ref, propsResolver))
 						.collect(Collectors.toList());
 	}
 
-	private ModularClassPath resolveDependency(String dep) {
-		DependencyResolver resolver = new DependencyResolver().addDependency(dep);
-		updateDependencyResolver(resolver);
-		return resolver.resolve();
-	}
-
 	public Code forResource(String resource) {
 		ResourceRef resourceRef = resolveChecked(getResourceResolver(), resource);
-
-		if (resourceRef instanceof AliasResourceResolver.AliasedResourceRef) {
-			// The resource we found was obtained from an alias which might
-			// contain extra options that need to be taken into account
-			// when running the code
-			Alias alias = ((AliasResourceResolver.AliasedResourceRef) resourceRef).getAlias();
-			if (getArguments() == null || getArguments().isEmpty()) {
-				setArguments(alias.arguments);
-			}
-			if (getJavaOptions() == null || getJavaOptions().isEmpty()) {
-				setJavaOptions(alias.javaOptions);
-			}
-			if (getAdditionalSources() == null || getAdditionalSources().isEmpty()) {
-				setAdditionalSources(alias.sources);
-			}
-			if (getAdditionalDependencies() == null || getAdditionalDependencies().isEmpty()) {
-				setAdditionalDependencies(alias.dependencies);
-			}
-			if (getAdditionalRepositories() == null || getAdditionalRepositories().isEmpty()) {
-				setAdditionalRepositories(alias.repositories);
-			}
-			if (getAdditionalClasspaths() == null || getAdditionalClasspaths().isEmpty()) {
-				setAdditionalClasspaths(alias.classpaths);
-			}
-			if (getProperties() == null || getProperties().isEmpty()) {
-				setProperties(alias.properties);
-			}
-			if (getJavaVersion() == null) {
-				setJavaVersion(alias.javaVersion);
-			}
-			if (getMainClass() == null) {
-				setMainClass(alias.mainClass);
-			}
-			setAlias(alias);
-		}
-
-		// note script file must be not null at this point
 		setOriginalRef(resource);
 		return forResourceRef(resourceRef);
 	}
@@ -525,7 +480,58 @@ public class RunContext {
 
 	private ResourceResolver getResourceResolver() {
 		Catalog catalog = getCatalog() != null ? Catalog.get(getCatalog().toPath()) : null;
-		return ResourceResolver.forScripts(this::resolveDependency, catalog);
+		return new AliasResourceResolver(catalog, this::getAliasResourceResolver);
 	}
 
+	private ResourceResolver getAliasResourceResolver(Alias alias) {
+		if (alias != null) {
+			updateFromAlias(alias);
+		}
+		return new CombinedResourceResolver(
+				new RenamingScriptResourceResolver(),
+				new LiteralScriptResourceResolver(),
+				new RemoteResourceResolver(false),
+				new ClasspathResourceResolver(),
+				new GavResourceResolver(getDependencyResolver(alias)),
+				new FileResourceResolver());
+	}
+
+	private Function<String, ModularClassPath> getDependencyResolver(Alias alias) {
+		return dep -> {
+			DependencyResolver resolver = new DependencyResolver().addDependency(dep);
+			updateDependencyResolver(resolver);
+			return resolver.resolve();
+		};
+	}
+
+	private void updateFromAlias(Alias alias) {
+		if (getArguments() == null || getArguments().isEmpty()) {
+			setArguments(alias.arguments);
+		}
+		if (getJavaOptions() == null || getJavaOptions().isEmpty()) {
+			setJavaOptions(alias.javaOptions);
+		}
+		if (getAdditionalSources() == null || getAdditionalSources().isEmpty()) {
+			setAdditionalSources(alias.sources);
+		}
+		if (getAdditionalDependencies() == null || getAdditionalDependencies().isEmpty()) {
+			setAdditionalDependencies(alias.dependencies);
+		}
+		if (getAdditionalRepositories() == null || getAdditionalRepositories().isEmpty()) {
+			setAdditionalRepositories(alias.repositories);
+		}
+		if (getAdditionalClasspaths() == null || getAdditionalClasspaths().isEmpty()) {
+			setAdditionalClasspaths(alias.classpaths);
+		}
+		if (getProperties() == null || getProperties().isEmpty()) {
+			setProperties(alias.properties);
+		}
+		if (getJavaVersion() == null) {
+			setJavaVersion(alias.javaVersion);
+		}
+		if (getMainClass() == null) {
+			setMainClass(alias.mainClass);
+		}
+		setAlias(alias);
+	}
 }

--- a/src/main/java/dev/jbang/source/resolvers/AliasResourceResolver.java
+++ b/src/main/java/dev/jbang/source/resolvers/AliasResourceResolver.java
@@ -2,6 +2,7 @@ package dev.jbang.source.resolvers;
 
 import java.io.File;
 import java.util.Objects;
+import java.util.function.Function;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
@@ -15,25 +16,29 @@ public class AliasResourceResolver implements ResourceResolver {
 	@Nullable
 	private final Catalog catalog;
 	@Nonnull
-	private final ResourceResolver resolver;
+	private final Function<Alias, ResourceResolver> resolverFactory;
 
-	public AliasResourceResolver(@Nullable Catalog catalog, @Nonnull ResourceResolver resolver) {
+	public AliasResourceResolver(@Nullable Catalog catalog,
+			@Nonnull Function<Alias, ResourceResolver> resolverFactory) {
 		this.catalog = catalog;
-		this.resolver = resolver;
+		this.resolverFactory = resolverFactory;
 	}
 
 	@Override
 	public String description() {
-		return String.format("Alias %s from catalog %s", resolver.description(),
-				Objects.toString(catalog.getScriptBase(), "<none>"));
+		return String.format("Alias resolver from catalog %s using %s",
+				Objects.toString(catalog.getScriptBase(), "<none>"),
+				resolverFactory.apply(null).description());
 	}
 
 	@Override
 	public ResourceRef resolve(String resource, boolean trusted) {
+		ResourceResolver resolver = resolverFactory.apply(null);
 		ResourceRef ref = resolver.resolve(resource, trusted);
 		if (ref == null) {
 			Alias alias = (catalog != null) ? Alias.get(catalog, resource) : Alias.get(resource);
 			if (alias != null) {
+				resolver = resolverFactory.apply(alias);
 				ResourceRef aliasRef = resolver.resolve(alias.resolve(), trusted);
 				if (aliasRef == null) {
 					throw new IllegalArgumentException(

--- a/src/test/java/dev/jbang/cli/TestAlias.java
+++ b/src/test/java/dev/jbang/cli/TestAlias.java
@@ -3,8 +3,11 @@ package dev.jbang.cli;
 import static dev.jbang.util.TestUtil.clearSettingsCaches;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.*;
+import static org.junit.jupiter.api.Assertions.fail;
 
 import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -92,8 +95,7 @@ public class TestAlias extends BaseTest {
 		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)), is(false));
 		JBang jbang = new JBang();
 		new CommandLine(jbang).execute("alias", "add", "-f", cwd.toString(), "--name=name", testFile.toString());
-		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)),
-				is(true));
+		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)), is(true));
 		Alias alias = Alias.get("name");
 		assertThat(alias.scriptRef, is("test.java"));
 	}
@@ -187,8 +189,7 @@ public class TestAlias extends BaseTest {
 		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)), is(false));
 		JBang jbang = new JBang();
 		new CommandLine(jbang).execute("alias", "add", "-f", cwd.toString(), "--name=name", testFile.toString());
-		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)),
-				is(true));
+		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)), is(true));
 		Alias name = Alias.get("name");
 		assertThat(name.scriptRef, is(Paths.get("sub/test.java").toString()));
 	}
@@ -202,8 +203,7 @@ public class TestAlias extends BaseTest {
 		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)), is(false));
 		JBang jbang = new JBang();
 		new CommandLine(jbang).execute("alias", "add", "-f", cwd.toString(), "--name=name", testFile.toString());
-		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)),
-				is(true));
+		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)), is(true));
 		Alias name = Alias.get("name");
 		assertThat(name.scriptRef, is("test.java"));
 		assertThat(name.description, is("Description of the script inside the script"));
@@ -219,8 +219,7 @@ public class TestAlias extends BaseTest {
 		JBang jbang = new JBang();
 		new CommandLine(jbang).execute("alias", "add", "-f", cwd.toString(), "--name=name",
 				"-d", "Description of the script in arguments", testFile.toString());
-		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)),
-				is(true));
+		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)), is(true));
 		Alias name = Alias.get("name");
 		assertThat(name.scriptRef, is("test.java"));
 		// The argument has more precedance than the description tag in the script
@@ -238,8 +237,7 @@ public class TestAlias extends BaseTest {
 		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)), is(false));
 		JBang jbang = new JBang();
 		new CommandLine(jbang).execute("alias", "add", "-f", cwd.toString(), "--name=name", testFile.toString());
-		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)),
-				is(true));
+		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)), is(true));
 		Alias name = Alias.get("name");
 		assertThat(name.scriptRef, is("test.java"));
 		assertThat(name.description,
@@ -254,8 +252,7 @@ public class TestAlias extends BaseTest {
 		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)), is(false));
 		JBang jbang = new JBang();
 		new CommandLine(jbang).execute("alias", "add", "-f", cwd.toString(), "--name=name", testFile.toString());
-		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)),
-				is(true));
+		assertThat(Files.isRegularFile(Paths.get(cwd.toString(), Catalog.JBANG_CATALOG_JSON)), is(true));
 		Alias name = Alias.get("name");
 		assertThat(name.scriptRef, is("test.java"));
 		assertThat(name.description,
@@ -290,6 +287,24 @@ public class TestAlias extends BaseTest {
 		Alias name = Alias.get("name");
 		assertThat(one.scriptRef, is("http://dummy"));
 		assertThat(name.scriptRef, is("test.java"));
+	}
+
+	@Test
+	void testAddWithRepos() throws IOException {
+		String jar = "dummygroup:dummyart:0.1";
+		CommandLine.ParseResult pr = JBang	.getCommandLine()
+											.parseArgs("alias", "add", "--name=aliaswithrepo", "--repos",
+													"http://dummyrepo", jar);
+		AliasAdd add = (AliasAdd) pr.subcommand().subcommand().commandSpec().userObject();
+		try {
+			add.doCall();
+			fail("Should have thrown exception");
+		} catch (ExitException ex) {
+			StringWriter sw = new StringWriter();
+			ex.printStackTrace(new PrintWriter(sw));
+			assertThat(sw.toString(), containsString(
+					"Error transferring file: dummyrepo from http://dummyrepo/dummygroup/dummyart/0.1/dummyart-0.1.pom"));
+		}
 	}
 
 	@Test

--- a/src/test/java/dev/jbang/cli/TestRun.java
+++ b/src/test/java/dev/jbang/cli/TestRun.java
@@ -493,6 +493,37 @@ public class TestRun extends BaseTest {
 	}
 
 	@Test
+	void testAliasWithRepo(@TempDir File output) throws IOException {
+		final String aliases = "{\n" +
+				"  \"aliases\": {\n" +
+				"    \"aliaswithrepo\": {\n" +
+				"      \"script-ref\": \"dummygroup:dummyart:0.1\",\n" +
+				"      \"repositories\": [ \"http://dummyrepo\" ]\n" +
+				"    }\n" +
+				"  }\n" +
+				"}";
+
+		environmentVariables.set("JBANG_DIR", jbangTempDir.toString());
+		Files.write(jbangTempDir.resolve(Catalog.JBANG_CATALOG_JSON), aliases.getBytes());
+
+		CommandLine.ParseResult pr = JBang.getCommandLine().parseArgs("run", "aliaswithrepo");
+
+		Run run = (Run) pr.subcommand().commandSpec().userObject();
+
+		RunContext ctx = run.getRunContext();
+
+		try {
+			ctx.forResource("aliaswithrepo");
+			fail("Should have thrown exception");
+		} catch (ExitException ex) {
+			StringWriter sw = new StringWriter();
+			ex.printStackTrace(new PrintWriter(sw));
+			assertThat(sw.toString(), containsString(
+					"Error transferring file: dummyrepo from http://dummyrepo/dummygroup/dummyart/0.1/dummyart-0.1.pom"));
+		}
+	}
+
+	@Test
 	void testHelloWorldShellNoExit() throws IOException {
 
 		environmentVariables.clear("JAVA_HOME");
@@ -1890,6 +1921,7 @@ public class TestRun extends BaseTest {
 
 		try {
 			ss.builder(ctx).build();
+			fail("Should have thrown exception");
 		} catch (ExitException ex) {
 			StringWriter sw = new StringWriter();
 			ex.printStackTrace(new PrintWriter(sw));
@@ -1912,6 +1944,7 @@ public class TestRun extends BaseTest {
 
 		try {
 			ctx.forResource(jar);
+			fail("Should have thrown exception");
 		} catch (ExitException ex) {
 			StringWriter sw = new StringWriter();
 			ex.printStackTrace(new PrintWriter(sw));
@@ -1935,6 +1968,7 @@ public class TestRun extends BaseTest {
 
 		try {
 			ctx.forResource(jar);
+			fail("Should have thrown exception");
 		} catch (ExitException ex) {
 			StringWriter sw = new StringWriter();
 			ex.printStackTrace(new PrintWriter(sw));
@@ -1951,6 +1985,7 @@ public class TestRun extends BaseTest {
 		RunContext ctx = run.getRunContext();
 		try {
 			SourceSet ss = (SourceSet) ctx.forResourceRef(LiteralScriptResourceResolver.stringToResourceRef(null, ""));
+			fail("Should have thrown exception");
 		} catch (ExitException ex) {
 			StringWriter sw = new StringWriter();
 			ex.printStackTrace(new PrintWriter(sw));
@@ -1981,6 +2016,7 @@ public class TestRun extends BaseTest {
 		code = run.prepareArtifacts(code, ctx);
 		try {
 			new JarCmdGenerator(code, ctx).generate();
+			fail("Should have thrown exception");
 		} catch (ExitException e) {
 			StringWriter sw = new StringWriter();
 			e.printStackTrace(new PrintWriter(sw));

